app/src/main/java/com/kickstarter/viewmodels/DiscoveryFragmentViewModel.java

```diff
@@ -301,4 +181,87 @@ private void saveLastSeenActivityId(final @Nullable Activity activity) {
+  private final PublishSubject<Void> nextPage = PublishSubject.create();
+  private final PublishSubject<DiscoveryParams> paramsFromActivity = PublishSubject.create();
+  private final PublishSubject<List<Category>> rootCategories = PublishSubject.create();
+
+  private final BehaviorSubject<Activity> activity = BehaviorSubject.create();
+  private final BehaviorSubject<List<Project>> projects = BehaviorSubject.create();
+  private final Observable<Boolean> showActivityFeed;
+  private final Observable<Activity> showActivityUpdate;
+  private final Observable<Boolean> showLoginTout;
+  private final PublishSubject<Pair<Project, RefTag>> showProject = PublishSubject.create();
+  private final BehaviorSubject<Boolean> shouldShowOnboardingView = BehaviorSubject.create();
+
+  public final DiscoveryFragmentViewModelInputs inputs = this;
+  public final DiscoveryFragmentViewModelOutputs outputs = this;
+
+  @Override public void activitySampleFriendBackingViewHolderProjectClicked(final @NonNull ActivitySampleFriendBackingViewHolder viewHolder,
+    final @NonNull Project project) {
+    activitySampleProjectClick.onNext(project);
+  }
+  @Override public void activitySampleFriendBackingViewHolderSeeActivityClicked(final @NonNull ActivitySampleFriendBackingViewHolder viewHolder) {
+    activityClick.onNext(true);
+  }
+  @Override public void activitySampleFriendFollowViewHolderSeeActivityClicked(final @NonNull ActivitySampleFriendFollowViewHolder viewHolder) {
+    activityClick.onNext(true);
+  }
+  @Override public void activitySampleProjectViewHolderProjectClicked(final @NonNull ActivitySampleProjectViewHolder viewHolder,
+    final @NonNull Project project) {
+    activitySampleProjectClick.onNext(project);
+  }
+  @Override public void activitySampleProjectViewHolderSeeActivityClicked(final @NonNull ActivitySampleProjectViewHolder viewHolder) {
+    activityClick.onNext(true);
+  }
+  @Override public void activitySampleProjectViewHolderUpdateClicked(final @NonNull ActivitySampleProjectViewHolder viewHolder,
+    final @NonNull Activity activity) {
+    activityUpdateClick.onNext(activity);
+  }
+  @Override public void rootCategories(final @NonNull List<Category> rootCategories) {
+    this.rootCategories.onNext(rootCategories);
+  }
+  @Override public void clearPage() {
+    clearPage.onNext(null);
+  }
+  @Override public void discoveryOnboardingViewHolderLoginToutClick(final @NonNull DiscoveryOnboardingViewHolder viewHolder) {
+    discoveryOnboardingLoginToutClick.onNext(true);
+  }
+  @Override public void nextPage() {
+    nextPage.onNext(null);
+  }
+  @Override public void paramsFromActivity(final @NonNull DiscoveryParams params) {
+    paramsFromActivity.onNext(params);
+  }
+  @Override public void projectCardViewHolderClick(final @NonNull ProjectCardViewHolder viewHolder, final Project project) {
+    clickProject.onNext(project);
+  }
+
+  @Override public Observable<Activity> activity() {
```

Comment:
I think all the outputs could also have non null annotations on the return type.